# Tutorials

A collection of Jupyter notebooks for using [OpenMendel](https://openmendel.github.io) packages. 

* PLINK binary data input, summary statistics, filtering, and visualization. \[[ipynb](./SnpArrays/genedata.ipynb)\]

* Kinship calculation and comparison. \[[ipynb](./Kinship/KinshipTutorial.ipynb)\]

* Population GWAS. \[[ipynb](./MendelGwas/GWASNotebook.ipynb)\]

* Iterative hard thresholding (IHT) for GWAS. \[[ipynb](./IHT/Mendel_IHT_tutorial.ipynb)\]

* Heritability estimation. \[[ipynb](./Heritability/HERITABILITY-VCexample.ipynb)\]

* GWAS based on linear mixed model. \[[ipynb](./LmmGwas/demo_LmmGWAS.ipynb)\]

* SNP-set testing. \[[ipynb](./VCTest/demo_exactVC.ipynb)\]

## Citation

If you use [OpenMendel](https://openmendel.github.io) analysis packages in your research, please cite the following reference in the resulting publications:

Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.

<!--- ## Contributing
We welcome contributions to this Open Source project. To contribute, follow this procedure ... --->

## Acknowledgments

This project is supported by the National Institutes of Health under NIGMS awards R01GM053275 and R25GM103774 and NHGRI award R01HG006139.